File: 557. Reverse Words in a String III/557.go
// 557. Reverse Words in a String III solution in GoLang
// https://leetcode.com/problems/reverse-words-in-a-string-iii/description/

package leetcode

func reverseWords(s string) string {
    words := strings.Split(s, " ")
    var result string

    for i := 0; i < len(words); i++ {
        if len(result) > 0 {
            result = result + " " + reverse(words[i])
        } else {
            result = result + reverse(words[i])
        }
    }

    return result
}

func reverse(word string) string {
    word_array := strings.Split(word, "");

    left, right := 0, len(word_array) - 1

    for left < right {
        temp := word_array[left]
        word_array[left] = word_array[right]
        word_array[right] = temp
        left++
        right--
    }

    return strings.Join(word_array, "")
}

File: 557. Reverse Words in a String III/557.java
// 557. Reverse Words in a String III solution in Java
// https://leetcode.com/problems/reverse-words-in-a-string-iii/description/

class Solution {
    public String reverseWords(String s) {
        List<String> words = Arrays.asList(s.split(" "));
        StringBuilder result = new StringBuilder();

        for (int i = 0; i < words.size(); i++) {
            if (result.length() == 0) {
                result.append(reverse(words.get(i)));
            } else {
                result.append(" ");
                result.append(reverse(words.get(i)));
            }
        }

        return result.toString();
    }

    private String reverse(String word) {
        char[] word_array = word.toCharArray();

        int left = 0;
        int right = word_array.length - 1;
        while (left < right) {
            char temp = word_array[left];
            word_array[left] = word_array[right];
            word_array[right] = temp;
            left++;
            right--;
        }

        return new String(word_array);
    }
}